File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/typescript/metadata_deadlines_cancellation_6.ts
import * as grpc from '@grpc/grpc-js';

// ===== CLIENT: attach a deadline + metadata =====
const deadline = new Date();
deadline.setSeconds(deadline.getSeconds() + 1); // absolute deadline

const meta = new grpc.Metadata();
const token = "my-token";
const reqID = "12345";
meta.add('authorization', 'Bearer ' + token);   // auth travels as metadata, sec 16
meta.add('x-request-id', reqID);              // trace id for correlation, like sec 15 of the layers manual

const client = new UserServiceClient('localhost:50051', grpc.credentials.createInsecure());
const req = new GetUserRequest();
req.setId("u42");

client.getUser(req, meta, { deadline }, (err: grpc.ServiceError, resp: any) => {
    if (err) {
        if (err.code === grpc.status.DEADLINE_EXCEEDED) {
            console.log("call timed out"); // a network-shaped failure a local call never had
        }
    }
});

// ===== SERVER: read metadata, respect the inherited deadline =====
class UserServer {
    getUser(
        call: grpc.ServerUnaryCall<GetUserRequest, GetUserResponse>,
        callback: grpc.sendUnaryData<GetUserResponse>
    ) {
        const md = call.metadata;
        const auth = md.get('authorization'); // verify token here (or in an interceptor, sec 15)

        // The gRPC Node library handles cancellation automatically. If the client 
        // cancels or times out, the call will emit a 'cancelled' event.
        if (call.cancelled) {
            return callback({ code: grpc.status.CANCELLED, message: 'Cancelled', name: 'Error' }, null);
        }
        
        call.on('cancelled', () => {
            // pass it straight to the DB driver so slow work is abandoned automatically
        });

        this.lookup(call.request.getId(), (err, user) => {
            if (call.cancelled) return;
            if (err) return callback(err, null);
            const resp = new GetUserResponse();
            resp.setUser(user);
            callback(null, resp);
        });
    }

    lookup(id: string, cb: (err: any, u: User) => void) { cb(null, new User()); }
}

// Dummy classes
class UserServiceClient { constructor(a:any,b:any){} getUser(a:any,b:any,c:any,d:any){} }
class GetUserRequest { setId(i:string){} getId(){ return ""; } }
class GetUserResponse { setUser(u:any){} }
class User {}

File: 13.gRPC-and-Inter-Service-Communication-for-Microservices/code/rust/metadata_deadlines_cancellation_6.rs
use tonic::{Request, Response, Status, metadata::MetadataValue};
use std::time::Duration;

// ===== CLIENT: attach a deadline + metadata =====
async fn call_with_meta(mut client: AnyServiceClient) {
    let mut req = Request::new(GetUserRequest { id: "u42".to_string() });
    
    // absolute deadline
    req.set_timeout(Duration::from_secs(1));
    
    let token = "my-token";
    let req_id = "12345";
    
    // auth travels as metadata, sec 16
    let auth_val: MetadataValue<_> = format!("Bearer {}", token).parse().unwrap();
    req.metadata_mut().insert("authorization", auth_val);
    
    // trace id for correlation, like sec 15 of the layers manual
    req.metadata_mut().insert("x-request-id", req_id.parse().unwrap());
    
    match client.get_user(req).await {
        Err(e) if e.code() == tonic::Code::DeadlineExceeded => {
            println!("call timed out"); // a network-shaped failure a local call never had
        }
        _ => {}
    }
}

// ===== SERVER: read metadata, respect the inherited deadline =====
#[tonic::async_trait]
impl MyServer {
    async fn get_user(&self, request: Request<GetUserRequest>) -> Result<Response<GetUserResponse>, Status> {
        let md = request.metadata();
        let auth = md.get("authorization"); // verify token here (or in an interceptor, sec 15)

        // In Tonic/Tokio, cancellation is handled by dropping the future.
        // If the client disconnects or times out, the `get_user` future is dropped.
        // pass it straight to the DB driver so slow work is abandoned automatically.

        let _ = auth;
        let id = request.into_inner().id;
        self.lookup(&id).await
    }
}

// Dummy structs
pub struct GetUserRequest { pub id: String }
pub struct GetUserResponse {}
pub struct AnyServiceClient;
impl AnyServiceClient { async fn get_user(&mut self, req: Request<GetUserRequest>) -> Result<Response<GetUserResponse>, Status> { unimplemented!() } }
pub struct MyServer;
impl MyServer { async fn lookup(&self, id: &str) -> Result<Response<GetUserResponse>, Status> { unimplemented!() } }
